packages/stateManager/index.ts

```diff
@@ -1,104 +1,636 @@
+      const ut = event.updateTag;
+      const tagId = bytesToHex(ut.tagId);
 
-class ItemManager extends PublicObjectManager<Item> {
-  constructor(db: Store<Item>, client: RelayClient) {
-    super(db, client);
-  }
-  // Process the events; if the event modifies one the shops items this updates the store and emits an event
-  // returns a bool to indicate whether the event was processed or not
-  async _processEvent(event: schema.ShopEvents): Promise<boolean> {
-    if (typeof event === schema.ShopEvents.UpdateItem) {
-      // load the item from the store
-      const item = await this.db.get(event.eventId);
-      Object.assign(item, event);
-      await this.db.put(event.eventId, item);
-      this.emit("update", item);
-      return true;
-    } else {
-      return false;
+      if (ut.addItemId) {
+        const itemId = bytesToHex(ut.addItemId);
+        const item = await this.store.get(itemId);
+        item.tags.push(tagId);
+        await this.store.put(itemId, item);
+        this.emit("addItemId", item);
+        return;
+      }
+      if (ut.removeItemId) {
+        const itemId = bytesToHex(ut.removeItemId);
+        const item = await this.store.get(itemId);
+        // remove `tagId` from item.tags array
+        item.tags = [...item.tags.filter((id: `0x${string}`) => id !== tagId)];
+        await this.store.put(itemId, item);
+        this.emit("removeItemId", item);
+        return;
+      }
     }
   }
-  create(item: Item) {
-    return Promise.all([
-      this.db.put(item.id, item),
-      this.client.createItem(item),
-    ]);
+
+  async create(item: Partial<Item>): Promise<Item> {
+    const eventId = await this.client.createItem({
+      price: item.price,
+      metadata: convertToBuffer(item.metadata),
+    });
+
+    return new Promise((resolve, reject) => {
+      const remove = () => this.removeListener("createItem", onCreate);
+      function onCreate(create: Item) {
+        if (create.id === bytesToHex(eventId)) {
+          resolve(create);
+          remove();
+        }
+      }
+      this.on("createItem", onCreate);
+    });
   }
+  //update argument passed here will only contain the fields to update.
+  async update(update: Partial<Item>): Promise<Item> {
+    //ui object to be passed to the network with converted network data types
+    const ui: schema.IUpdateItem = {
+      itemId: hexToBytes(update.id!),
+    };
+    if (update.price) {
+      ui.price = update.price;
+    }
+    if (update.metadata) {
+      ui.metadata = convertToBuffer(update.metadata);
+    }
+    await this.client.updateItem(ui);
 
-  update(item: Item) {
-    return Promise.all([
-      // need to get the item and merge it before storing
-      this.db.put(item.id, item),
-      this.client.createItem(item),
-    ]);
+    return new Promise((resolve, reject) => {
+      const remove = () => this.removeListener("updateItem", onUpdate);
+      function onUpdate(item: Item) {
+        if (update.id === item.id) {
```

Comment:
this is deceptively working but incorrect.

you need to be careful here to not unblock i.e. the UI calling `.update()` via a network request also updating the same object. In short, this needs to use the `eventId` not the `update.itemId`.

Just to elaborate the bug a bit:

- the UI could call `stateManager.update(itemId=23, price=$somePrice)`, the network client will create an eventId=123 for the update that goes out to the network momentarily.
- another client/clerk might have updated that same item just before
- the stateManager receives `updateItem(eventId=666, itemId=23, title=$newTitle)`

now, if you listen for the itemId instead of the eventId, you unblock for the wrong update and have a gnarly concurrency bug on your hand.